File: kernel/common/src/process/process.c
// NebulaOS - Process Management
// ================================
//
// Process control blocks and context switching

#include "../../common/include/process.h"
#include "../../common/include/nebula.h"
#include "../../common/include/stdint.h"
#include "../../common/include/scheduler.h"
#include "../../lib/include/string.h"
#include "../../kernel/common/include/idt.h"
#include "../../kernel/common/include/memory.h"

static process_t process_table[MAX_PROCESSES];
static uint32_t next_pid = 1;
static process_t* current_process = NULL;
static bool process_initialized = false;

void process_init(void) {
    memset(process_table, 0, sizeof(process_table));
    
    for (int i = 0; i < MAX_PROCESSES; i++) {
        process_table[i].pid = 0;
        process_table[i].state = PROC_ZOMBIE;
        process_table[i].next = NULL;
    }
    
    current_process = NULL;
    process_initialized = true;
}

int process_create(void* entry, uint32_t stack_size) {
    if (!process_initialized) return -1;
    if (!entry || stack_size == 0) return -1;
    
    for (int i = 0; i < MAX_PROCESSES; i++) {
        if (process_table[i].state == PROC_ZOMBIE) {
            process_t* proc = &process_table[i];
            
            proc->pid = next_pid++;
            proc->state = PROC_READY;
            proc->entry_point = entry;
            proc->stack_size = ALIGN_UP(stack_size, 16);
            proc->stack_base = (uint64_t)malloc(proc->stack_size) + proc->stack_size;
            
            memset(&proc->context, 0, sizeof(cpu_context_t));
            proc->context.rip = (uint64_t)entry;
            proc->context.rsp = proc->stack_base;
            proc->context.cs = 0x08;
            proc->context.ss = 0x10;
            proc->context.rflags = 0x202;
            proc->context.rdi = 0;
            proc->context.rsi = 0;
            proc->context.rdx = 0;
            proc->context.rcx = 0;
            proc->context.rbx = 0;
            proc->context.rax = 0;
            proc->context.rbp = proc->stack_base;
            
            for (int f = 0; f < 8; f++) {
                proc->fd_table[f] = -1;
            }
            
            proc->next = NULL;
            
            if (!current_process) {
                current_process = proc;
            }
            
            scheduler_add(proc->pid);
            
            return (int)proc->pid;
        }
    }
    
    return -1;
}

void process_destroy(uint32_t pid) {
    if (!process_initialized) return;
    
    for (int i = 0; i < MAX_PROCESSES; i++) {
        if (process_table[i].pid == pid) {
            process_table[i].state = PROC_ZOMBIE;
            if (current_process == &process_table[i]) {
                current_process = NULL;
            }
            return;
        }
    }
}

process_t* process_get_current(void) {
    return current_process;
}

void process_set_current(uint32_t pid) {
    for (int i = 0; i < MAX_PROCESSES; i++) {
        if (process_table[i].pid == pid) {
            current_process = &process_table[i];
            return;
        }
    }
}

void process_context_switch(cpu_context_t* regs) {
    if (!current_process) return;
    
    current_process->state = PROC_READY;
    memcpy(&current_process->context, regs, sizeof(cpu_context_t));
}

process_t* process_next(void) {
    uint32_t current_pid = current_process ? current_process->pid : 0;
    int start = -1;
    
    for (int i = 0; i < MAX_PROCESSES; i++) {
        if (process_table[i].pid == current_pid) {
            start = i;
            break;
        }
    }
    
    if (start < 0) start = 0;
    
    for (int i = 1; i <= MAX_PROCESSES; i++) {
        int idx = (start + i) % MAX_PROCESSES;
        if (process_table[idx].state == PROC_READY && process_table[idx].pid != current_pid) {
            return &process_table[idx];
        }
    }
    
    return current_process;
}

uint32_t process_get_count(void) {
    uint32_t count = 0;
    for (int i = 0; i < MAX_PROCESSES; i++) {
        if (process_table[i].state != PROC_ZOMBIE) {
            count++;
        }
    }
    return count;
}

process_state_t process_get_state(uint32_t index) {
    if (index < MAX_PROCESSES) {
        return process_table[index].state;
    }
    return PROC_ZOMBIE;
}

uint32_t process_get_pid(uint32_t index) {
    if (index < MAX_PROCESSES && process_table[index].state != PROC_ZOMBIE) {
        return process_table[index].pid;
    }
    return 0;
}

void* process_get_entry_point(uint32_t index) {
    if (index < MAX_PROCESSES && process_table[index].state != PROC_ZOMBIE) {
        return process_table[index].entry_point;
    }
    return NULL;
}

File: common/src/process.rs
use crate::stdint::*;

pub const MAX_PROCESSES: usize = 16;

#[repr(C)]
#[derive(Clone, Copy)]
pub struct CpuContext {
    pub r15: u64, pub r14: u64, pub r13: u64, pub r12: u64,
    pub r11: u64, pub r10: u64, pub r9: u64, pub r8: u64,
    pub rbp: u64, pub rdi: u64, pub rsi: u64, pub rdx: u64,
    pub rcx: u64, pub rbx: u64, pub rax: u64,
    pub rip: u64, pub cs: u64, pub rflags: u64, pub rsp: u64, pub ss: u64,
}

#[repr(C)]
#[derive(Clone, Copy)]
pub struct Process {
    pub pid: u32,
    pub state: u32,
    pub context: CpuContext,
    pub stack_base: usize,
    pub stack_size: u32,
    pub entry_point: *mut u8,
    pub fd_table: [i32; 8],
}

pub const PROC_READY: u32 = 0;
pub const PROC_RUNNING: u32 = 1;
pub const PROC_BLOCKED: u32 = 2;
pub const PROC_ZOMBIE: u32 = 3;

pub static mut PROCESS_TABLE: [Process; MAX_PROCESSES as usize] = [Process { pid: 0, state: 0, context: unsafe { core::mem::zeroed() }, stack_base: 0, stack_size: 0, entry_point: core::ptr::null_mut(), fd_table: [-1; 8] }; MAX_PROCESSES as usize];
static mut NEXT_PID: u32 = 1;
static mut CURRENT_PROCESS: *mut Process = core::ptr::null_mut();

pub unsafe fn process_init() {
    for i in 0..MAX_PROCESSES {
        (*PROCESS_TABLE.as_mut_ptr().add(i as usize)).pid = 0;
        (*PROCESS_TABLE.as_mut_ptr().add(i as usize)).state = PROC_ZOMBIE;
    }
    CURRENT_PROCESS = core::ptr::null_mut();
}

pub unsafe fn process_create(entry: *mut u8, stack_size: u32) -> i32 {
    for i in 0..MAX_PROCESSES {
        let proc = &mut *PROCESS_TABLE.as_mut_ptr().add(i as usize);
        if proc.state == PROC_ZOMBIE {
            proc.pid = NEXT_PID;
            NEXT_PID += 1;
            proc.state = PROC_READY;
            proc.entry_point = entry;
            proc.stack_size = (stack_size + 15) & !15;
            proc.stack_base = 0x10000000 + (i as usize) * 0x10000;
            proc.context = CpuContext {
                rip: entry as u64,
                rsp: proc.stack_base as u64,
                cs: 0x08,
                ss: 0x10,
                rflags: 0x202,
                ..unsafe { core::mem::zeroed() }
            };
            for f in 0..8 {
                proc.fd_table[f] = -1;
            }
            if CURRENT_PROCESS.is_null() {
                CURRENT_PROCESS = proc as *mut Process;
            }
            return proc.pid as i32;
        }
    }
    -1
}

pub unsafe fn process_get_current() -> *mut Process {
    CURRENT_PROCESS
}

pub unsafe fn process_set_current(pid: u32) {
    for i in 0..MAX_PROCESSES {
        if (*PROCESS_TABLE.as_ptr().add(i as usize)).pid == pid {
            CURRENT_PROCESS = PROCESS_TABLE.as_mut_ptr().add(i as usize);
            return;
        }
    }
}

pub unsafe fn process_destroy(pid: u32) {
    for i in 0..MAX_PROCESSES {
        if (*PROCESS_TABLE.as_ptr().add(i as usize)).pid == pid {
            (*PROCESS_TABLE.as_mut_ptr().add(i as usize)).state = PROC_ZOMBIE;
            if CURRENT_PROCESS == PROCESS_TABLE.as_ptr().add(i as usize) as *mut Process {
                CURRENT_PROCESS = core::ptr::null_mut();
            }
            return;
        }
    }
}
